pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -751,6 +754,11 @@ func GetResourceRequest(pod *v1.Pod) *schedulernodeinfo.Resource {
 		result.SetMaxResource(container.Resources.Requests)
 	}
 
+	// If Overhead is being utilized, add to the total requests for the pod
+	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
```

Comment:
Checking feature gates in the critical path of scheduling has shown to cause performance degradation. Feature gates should be read once and passed to the scheduler.

Reply:
Understood.  I am looking now for a pre-existing example, though nothing immediate comes up (looking at commit history and scheduler pkg).

Reply:
Here is an example: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/factory/factory.go#L290

Reply:
Thanks @bsalamat.  In the case of enableNonPreempting, it is relatively straight forward since the consumption of the flag is limited to within the genericScheduler itself.

In the case of enablePodOverhead, it is needed in three places (at least):
 - [node-info](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/nodeinfo/node_info.go#L591-L602): 
```golang
	// If Overhead is being utilized, add to the total requests for the pod
	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
		resPtr.Add(pod.Spec.Overhead)

		if _, found := pod.Spec.Overhead[v1.ResourceCPU]; found {
			non0CPU += pod.Spec.Overhead.Cpu().MilliValue()
		}

		if _, found := pod.Spec.Overhead[v1.ResourceMemory]; found {
			non0Mem += pod.Spec.Overhead.Memory().Value()
		}
	}
```
 - [predicate's GetResourceRequest](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/algorithm/predicates/predicates.go#L746-L763)
 - [priority's resource allocation requests](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/algorithm/priorities/resource_allocation.go#L94-L116)

The problem I have is finding the appropriate structure to stash the PodOverhead flag in.

## First attempt -- adding to NodeInfo structure:
NodeInfo is available for each area, but this doesn't work well, since: 1) feature gate is looked up per node, and 2) the GetMetadata for both priority and [predicates](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/algorithm/predicates/metadata.go#L155) calls into the applicable function where node specific information isn't available (ie, GetMetadata operates on a set of nodes)

## Second attempt -- add to the {Priority|Predicate}MetadataFactory

This isn't straightforward either, but may be next 'best' option. Modifications would be needed within the algorithmprovider for creating New{Priority|Predicate}MetadataFactory.

WDYT @bsalamat ? This match your expectation?

Reply:
Thanks for checking the options. You are right that this particular feature gate is harder to pass to various parts of the scheduler. Unfortunately, none of the possible options are sufficiently clean. So, let's keep them as they are. Since we check  `pod.Spec.Overhead != nil` before checking the feature gate, this will not have a performance impact for pods that don't use the feature.